### ISSUE TITLE
kubectl apply emits error when patching a ScheduledJob

### ISSUE THREAD

**Issue by vreon** [kubernetes/kubernetes#35149]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

Nope!

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

kubectl, apply, scheduledjob, batch/v1

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"clean", BuildDate:"2016-10-10T18:19:49Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"clean", BuildDate:"2016-10-10T18:13:36Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): `Debian GNU/Linux 8 (jessie)`
- **Kernel** (e.g. `uname -a`): `Linux ip-172-20-61-42 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-2 (2016-04-08) x86_64 GNU/Linux`
- **Install tools**: `kops`
- **Others**: n/a

**What happened**:

`kubectl apply` emits an error when applying changes to an existing `ScheduledJob`.

**What you expected to happen**:

`kubectl apply` should have successfully applied changes to the existing `ScheduledJob`.

**How to reproduce it** (as minimally and precisely as possible):

Given this `ScheduledJob` manifest:

``` yaml
# scheduledjob.yaml
# From http://kubernetes.io/docs/user-guide/scheduled-jobs/#creating-a-scheduled-job
apiVersion: batch/v2alpha1
kind: ScheduledJob
metadata:
  name: hello
spec:
  schedule: 0/1 * * * ?
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```

The first apply succeeds:

``` console
$ kubectl apply -f scheduledjob.yaml 
scheduledjob "hello" created
```

The second one fails:

``` console
$ kubectl apply -f scheduledjob.yaml
error: error when applying patch:

to:
&{0xc8200bd740 0xc8202f8ee0 sandbox hello scheduledjob.yaml &ScheduledJob{ObjectMeta:k8s_io_kubernetes_pkg_api_v1.ObjectMeta{Name:hello,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{kubectl.kubernetes.io/last-applied-configuration: ,},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:ScheduledJobSpec{Schedule:0/1 * * * ?,StartingDeadlineSeconds:nil,ConcurrencyPolicy:,Suspend:nil,JobTemplate:JobTemplateSpec{ObjectMeta:k8s_io_kubernetes_pkg_api_v1.ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:JobSpec{Parallelism:nil,Completions:nil,ActiveDeadlineSeconds:nil,Selector:nil,ManualSelector:nil,Template:k8s_io_kubernetes_pkg_api_v1.PodTemplateSpec{ObjectMeta:ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:PodSpec{Volumes:[],Containers:[{hello busybox [] [/bin/sh -c date; echo Hello from the Kubernetes cluster]  [] [] {map[] map[]} [] nil nil nil   nil false false false}],RestartPolicy:OnFailure,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,DNSPolicy:,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:nil,ImagePullSecrets:[],Hostname:,Subdomain:,},},},},},Status:ScheduledJobStatus{Active:[],LastScheduleTime:<nil>,},} &TypeMeta{Kind:,APIVersion:,} 12601558 false}
for: "scheduledjob.yaml": error when serializing current configuration from:
&TypeMeta{Kind:,APIVersion:,}
for: "scheduledjob.yaml": no kind "ScheduledJob" is registered for version "batch/v1"
```

**Anything else do we need to know**:
- `ScheduledJob` is enabled on the apiserver via `--runtime-config=batch/v2alpha1=true`
- first reported downstream at kubernetes/helm#1408

**Comment by vreon** [comment]
Found kubernetes/kubernetes#34010 while looking through the changelog; this is probably a duplicate of that issue. Will upgrade and confirm.

**Comment by vreon** [comment]
No dice; the same behavior is present in `1.4.3`.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.3", GitCommit:"4957b090e9a4f6a68b4a40375408fdc74a212260", GitTreeState:"clean", BuildDate:"2016-10-16T06:36:33Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.3", GitCommit:"4957b090e9a4f6a68b4a40375408fdc74a212260", GitTreeState:"clean", BuildDate:"2016-10-16T06:20:04Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by zreigz** [comment]
I can also reproduce it when I use code from master branch. I can take a look closer what cause the problem

**Comment by zreigz** [comment]
The problem starts here: https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/scheme.go#L515

The returned object looks like this:
```
unversioned.GroupVersionKind{
Group: batch,
Version: __internal,
Kind: ScheduledJob
}
```
Few lines later
```
gvk, ok := target.KindForGroupVersionKinds(kinds)
``` 

The method returns wrong value of GroupVersionKind because in method https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/group_version.go#L271 there are two possible `targets` and function `bestMatch` returns the first one but the second is the correct one.
But the reason is the previous returned `kinds []GroupVersionKind` which contains data:
```
{
Group: batch,
Version: __internal,
Kind: ScheduledJob
}
```
where Version field doesn't match to any `targets` because of strange Version value `__internal`

This is what is can see from debug. I am not expert in this area.
cc @soltysh @smarterclayton can you halp ?

**Comment by soltysh** [comment]
I'll look into it, I've solved similar problem with edit.

**Comment by soltysh** [comment]
Working on this right now...

**Comment by zreigz** [comment]
The main problem is that there are two versions of GroupVersionKind returned in `bestMatch` function. 
However ScheduledJob is registered only for v2alpha1 but it requires some additional check (or just workaround for this case). 
Because the file https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/group_version.go is in unversioned package there is circular dependency error when you want to add some new packages to make some check. In my opinion the best way is to move `group_version.go` to runtime package like it is suggested here: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/group_version.go#L264 but it affects a lot of files :(

**Comment by soltysh** [comment]
> where Version field doesn't match to any targets because of strange Version value __internal

That's an internal version marker, I'm wondering why it's internal here and not a specific version like it was in the case of edit.

**Comment by zreigz** [comment]
Yes you are right. I've noticed that later after further investigation

**Comment by soltysh** [comment]
I think we need to use the versioned object in apply, that's how it works in edit. Otherwise we'll always struggle with this problem. This worked ok with single versioned object, because it'll always returned the only available counterpart to `__internal`, but now that `batch` has two it's hard to guess which one to use, because `GroupVersions` knows nothing about kinds.

**Comment by zreigz** [comment]
Exactly. The problem is here:
`
    "/apis/batch/v1",
    "/apis/batch/v2alpha1",
`
GroupVersionKind with ScheduledJob is constructed based on `batch` group which has two versions. There is no possibility to check to which version ScheduledJob is registered

**Comment by zreigz** [comment]
I couldn't add any helpful logic to `group_version.go` because of circular dependency errors. I think this file should be in `runtime` package

**Comment by soltysh** [comment]
I have a working patch that modifies `kubectl apply` in such a way that it works against versioned object, instead of `__internal`, similarly to `kubectl edit`. PR will be opened shortly.

**Comment by JoeEnnever** [comment]
Hey we're running into this right now, is there any way to work around this bug until the fix is out? Aside from deleting and recreating the scheduledjob each time.

**Comment by soltysh** [comment]
@JoeEnnever `kubectl edit` works, it suffered similar problem quite some time ago but that's already fixed.

**Comment by pwittrock** [comment]
@soltysh Any updates on this?

**Comment by soltysh** [comment]
@pwittrock I'm stuck with it :disappointed: so I can't promise anything, but I'll try to get back to it.

**Comment by pwittrock** [comment]
@soltysh No problem, just trying to understand the state of the world

**Comment by pwittrock** [comment]
Let me know if you need help with any of it

**Comment by alex88** [comment]
Why closed? It's still an issue here, even after deleting the cronjob I get the same issue

**Comment by liggitt** [comment]
What client and server version do you have, what does your yaml file look like, and what is the output of the command. If possible, can you also include verbose output (--v=8), but redact any authorization headers

**Comment by alex88** [comment]
@liggitt just found out that it wasn't enough to delete the job in the jobs section of kubectl admin, I've deleted the cronjob via kubectl and it worked, you still need the other info?

**Comment by liggitt** [comment]
> you still need the other info?

depends on what version of kubectl you had and what error you were seeing. just want to make sure the issue is actually resolved.

**Comment by alex88** [comment]
@liggitt ok so

```
➜  ~ kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T07:31:07Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

the cronjob configuration is

```
apiVersion: batch/v2alpha1
kind: CronJob
metadata:
  name: cronjob
  namespace: project-test
spec:
  schedule: "*/2 * * * *"
  concurrencyPolicy: Forbid
  jobTemplate:
    spec:
      template:
        spec:
          restartPolicy: Never
          containers:
            - name: cronjob
              image: xxx.dkr.ecr.eu-west-1.amazonaws.com/project/web/backend
              command: ["/bin/run-console-command", "project:cronjob"]
              resources:
                requests:
                  memory: 512Mi
                limits:
                  memory: 512Mi
              volumeMounts:
                - mountPath: /var/www/project/app/config/environment
                  name: config
                - mountPath: /usr/local/etc/php/conf.d/environment
                  name: php-config
          volumes:
            - name: config
              configMap:
                name: symfony-config
                items:
                - key: environment
                  path: config.yml
            - name: php-config
              configMap:
                name: symfony-config
                items:
                - key: php_ini
                  path: config.ini
```

due an image error I got tons of errors since kube was continuously trying to start the containers. I went into the jobs page and deleted the cronjob job, then to restart it I've tried to run apply again and I got

```
error: error when applying patch:

to:
&{0xc4200dad80 0xc4203eed20 project-test cronjob php.yaml &CronJob{ObjectMeta:k8s_io_kubernetes_pkg_api_v1.ObjectMeta{Name:cronjob,GenerateName:,Namespace:project-test,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{kubectl.kubernetes.io/last-applied-configuration: ,},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:CronJobSpec{Schedule:*/2 * * * *,StartingDeadlineSeconds:nil,ConcurrencyPolicy:Forbid,Suspend:nil,JobTemplate:JobTemplateSpec{ObjectMeta:k8s_io_kubernetes_pkg_api_v1.ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:JobSpec{Parallelism:nil,Completions:nil,ActiveDeadlineSeconds:nil,Selector:nil,ManualSelector:nil,Template:k8s_io_kubernetes_pkg_api_v1.PodTemplateSpec{ObjectMeta:ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:PodSpec{Volumes:[{config {nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil ConfigMapVolumeSource{LocalObjectReference:LocalObjectReference{Name:symfony-config,},Items:[{environment config.yml <nil>}],DefaultMode:nil,} nil nil nil nil}} {php-config {nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil &ConfigMapVolumeSource{LocalObjectReference:LocalObjectReference{Name:symfony-config,},Items:[{php_ini config.ini <nil>}],DefaultMode:nil,} nil nil nil nil}}],Containers:[{cronjob 421458412047.dkr.ecr.eu-west-1.amazonaws.com/project/web/backend [/bin/run-console-command project:cronjob] []  [] [] {map[memory:{{536870912 0} {<nil>}  BinarySI}] map[memory:{{536870912 0} {<nil>}  BinarySI}]} [{config false /var/www/project/app/config/environment } {php-config false /usr/local/etc/php/conf.d/environment }] nil nil nil   nil false false false}],RestartPolicy:Never,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,DNSPolicy:,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:nil,ImagePullSecrets:[],Hostname:,Subdomain:,},},},},},Status:CronJobStatus{Active:[],LastScheduleTime:<nil>,},} &TypeMeta{Kind:,APIVersion:,} 1446604 false}
for: "php.yaml": error when serializing current configuration from:
&TypeMeta{Kind:,APIVersion:,}
for: "php.yaml": no kind "CronJob" is registered for version "batch/v1"
```

**Comment by liggitt** [comment]
I see. the issue is fixed in master, which will be in the 1.6 release.

**Comment by alex88** [comment]
Gotcha, thank you @liggitt
